### PR TITLE
rewrite write efficient queries

### PR DESCRIPTION
**Pull request by 0xBoxer**
Reworking the entire "write good queries" section to explain databases, parquet files and add lots of DuneSQL tips.

### CODE REVIEW

File: query-engine/writing-efficient-queries.mdx [hunk 1]
```diff
@@ -1,54 +1,192 @@
...
+
+2. **Row Group Level:** Once the appropriate Parquet file is identified, the engine will access the relevant row groups within the file, based on the query conditions. Again, it will first read the metadata of each row group to determine whether it might contain the data needed for the query. If the row group does not contain the data needed for the query, it will skip the row group and move on to the next one. This will further reduce the amount of data that needs to be read into memory, which will improve query performance.
+
+3. **Column Chunk Level:** Once the appropriate row group is identified, the system will access the relevant column chunks within the row group, based on the query conditions. The column chunks contain the actual data that is needed for the query. The database will only read the column chunks that contain the data needed for the query. It will not read the logical row - saving time and resources.
```
Comment (kasiafi) [inline query-engine/writing-efficient-queries.mdx:107]: > The database will only read the column chunks that contain the data needed for the query.

To showcase that you could change the example a bit:
```
SELECT <choose some columns instead of *> FROM ethereum.transactions
WHERE block_number = 14854616
AND hash = 0xce1f1a2dd0c10fcf9385d14bc92c686c210e4accf00a3fe7ec2b5db7a5499cff
```
and then say that only the column chunks for the required columns will be read.

File: query-engine/writing-efficient-queries.mdx [hunk 2]
```diff
@@ -1,54 +1,192 @@
...
+  to,
+  "from",
+  value,
+  block_date,
+  sqrt_price_x96,
+FROM
+  uniswap_v3_ethereum.Pair_evt_Swap swap
+  inner join ethereum.transactions t 
+    on swap.evt_tx_hash = t.hash
+    and swap.evt_block_number = t.block_number
```
Comment (kasiafi) [inline query-engine/writing-efficient-queries.mdx:149]: Yes! If users do this, it will be great!

File: query-engine/writing-efficient-queries.mdx (hunk 1 above)
Comment (kasiafi) [reply query-engine/writing-efficient-queries.mdx:107]: One more note: most of our data is partitioned on `block_date`. We should suggest filtering on `block_date` stronger than on `block_number`. Performance-wise, it is better to filter on partition column because it can remove a filter from the plan.

File: query-engine/writing-efficient-queries.mdx [hunk 3]
```diff
@@ -1,54 +1,192 @@
...
+where
+  swap.contract_address = 0x510100D5143e011Db24E2aa38abE85d73D5B2177
 ```
 
-By including the block_number column in the query, the engine can narrow down the search to a specific block, reducing the amount of data scanned and considerably speeding up the query.
+By including the `block_number` column in the join condition, the engine can quickly narrow down the search in `ethereum.transactions` to files and row groups that contain the `block_number` contained in a specific row in `uniswap_v3_ethereum.Pair_evt_Swap`. This significantly reduces the amount of data that needs to be read, thereby improving query performance.
+
+It also makes a difference whether we join `ethereum.transactions` on `uniswap_v3_ethereum.Pair_evt_Swap` or the other way around. The following query will be even faster:
```
Comment (kasiafi) [inline query-engine/writing-efficient-queries.mdx:156]: > join `ethereum.transactions` on `uniswap_v3_ethereum.Pair_evt_Swap`

I would reword to "join ... to ..." or "join ... with ..."

File: query-engine/writing-efficient-queries.mdx [hunk 4]
```diff
@@ -58,10 +196,19 @@ In addition to leveraging the columnar storage format and using sequentially ord
 
 6. **Avoid using DISTINCT when possible**: DISTINCT can be computationally expensive, especially on large datasets. If you can use GROUP BY or other aggregation methods to achieve the same result, it may improve query performance. Try using approx_distinct instead.
```
Comment (kasiafi) [inline query-engine/writing-efficient-queries.mdx:197]: > If you can use GROUP BY or other aggregation methods

DISTINCT is implemented as a GROUP BY, so it won't help if people do GROUP BY manually. The suggestion about approx_distinct is valid.

File: query-engine/writing-efficient-queries.mdx [hunk 5]
```diff
@@ -58,10 +196,19 @@ In addition to leveraging the columnar storage format and using sequentially ord
...
 
-7. **Optimize subqueries**: Subqueries can sometimes cause performance issues. Consider using Common Table Expressions (CTEs) or rewriting the query using JOINs to optimize subqueries.
+7. **Optimize subqueries**: Subqueries (Select * from (select * from table) as x) can sometimes cause performance issues. Consider using Common Table Expressions (CTEs) or rewriting the query using JOINs to optimize subqueries.
```
Comment (kasiafi) [inline query-engine/writing-efficient-queries.mdx:199]: CTEs are no different than in-line subqueries.

I would add a warning about CTE abuse: CTEs are inlined in each place where they're referenced (they're not precomputed) -- and so using CTEs multiple times, or combining CTEs in CTEs can lead to huge query plans.

File: query-engine/writing-efficient-queries.mdx [hunk 6]
```diff
@@ -58,10 +196,19 @@ In addition to leveraging the columnar storage format and using sequentially ord
...
 
 8. **Optimize data types**: Use appropriate data types for your columns, as it can improve query performance by reducing the amount of data processed. For example, varbinary operations are faster that varchar so be careful casting around too much.
 
-9. **Make sure to join the smaller table to the larger table**: This saves on memory use and will make your queries run faster. So you should be joining erc20 transfers onto ethereum transactions (as an example). 
+9. **Only order when necessary**: Ordering results can be computationally expensive. If you don't need ordered results, avoid using ORDER BY.
+
+10. **Use `UNION ALL` instead of `UNION`**: If you're combining the results of multiple queries, use `UNION ALL` instead of `UNION` to avoid the overhead of removing duplicate rows.
```
Comment (kasiafi) [inline query-engine/writing-efficient-queries.mdx:205]: You can put this point after (6) - they're both about DISTINCT. UNION is DISTINCT by default.

File: query-engine/writing-efficient-queries.mdx [hunk 7]
```diff
@@ -58,10 +196,19 @@ In addition to leveraging the columnar storage format and using sequentially ord
...
+
+12. **Use the `EXPLAIN` command**: Use the `EXPLAIN` command to understand how the query engine processes your query. This can help you identify potential performance bottlenecks and optimize your queries accordingly.
+
+13. **Check your execution statistics**: Click on "last run x seconds ago" to see the execution statistics for your query. This can help you identify potential performance bottlenecks and optimize your queries accordingly.
+
 
-By following these tips, you can write more efficient queries on DuneSQL with TrinoSQL and optimize the performance of your data processing tasks. Remember that DuneSQL's unique structure, such as the parquet file format and columnar storage, should be taken into account when optimizing your queries to fully benefit from the system's capabilities.
\ No newline at end of file
+By following these tips, you can write more efficient queries on DuneSQL with TrinoSQL and optimize the performance of your data processing tasks. Remember that DuneSQL's unique structure, such as the parquet file format and columnar storage, should be taken into account when optimizing your queries to fully benefit from the system's capabilities.
```
Comment (kasiafi) [inline query-engine/writing-efficient-queries.mdx:214]: Same comment about `TrinoSQL` :)

File: query-engine/writing-efficient-queries.mdx [hunk 8]
```diff
@@ -1,54 +1,192 @@
...
+  and t.block_number = swap.evt_block_number
+where
+  swap.contract_address = 0x510100D5143e011Db24E2aa38abE85d73D5B2177
+```
+
+<info> Using `block_date` is often better than using `block_number` or block_time as there is less distinct values to check against. Use `block_date` during joins where it is possible. </info>  
+
+This join order optimization is related to the size of the tables. **You should always join the smaller table on the bigger table**. In this case, `ethereum.transactions` is much bigger than `uniswap_v3_ethereum.Pair_evt_Swap`, so we should join `uniswap_v3_ethereum.Pair_evt_Swap` on `ethereum.transactions`. 
+
+Additionally, to enjoy these optimizations, **you will have to use inner joins**. Left joins and right joins will not benefit in the same way.
```
Comment (kasiafi) [inline query-engine/writing-efficient-queries.mdx:177]: Outer joins also benefit from the optimizations.

Inner joins are generally more lightweight though. Maybe suggest that the users should use inner joins whenever that's sufficient for their use case.

**Review by kasiafi — COMMENTED** [review]
Great read! I added some comments.

File: query-engine/writing-efficient-queries.mdx (hunk 1 above)
Comment (0xBoxer) [reply query-engine/writing-efficient-queries.mdx:107]: The problem is that we don't have `block_date` in all tables - I have been trying to lobby for it for a long time, but no success so far.

File: query-engine/writing-efficient-queries.mdx (hunk 2 above)
Comment (0xBoxer) [reply query-engine/writing-efficient-queries.mdx:144]: just FYI, when I started on this topic, Jonas did a bunch of testing and diving into query plans and found that this only works great if the range of distinct values that the engine has to join on is not very large. 
This might work for something like 5k rows(IIRC, check with Jonas if that number adds up), but afterward the engine goes back to brute forcing the hash joins. 

That's why having `block_date` would be so important, the range of distinct values would go down and the engine could do these joins much more efficiently.

**Comment by 0xBoxer** [comment]
This was a super helpful review, thank you so much @kasiafi 🙏
